[sparsity] Base sparsity level scheduler class

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* **#59770 [sparsity] Base sparsity level scheduler class**
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* #58705 [sparsity] Sparsity parametrization
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

Commit:
[sparsity] Base sparsity level scheduler class

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"




[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"




[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"




Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"




Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"




Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"




Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"




Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"




Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

raghuramank100 commented on torch/ao/sparsity/scheduler/base_scheduler.py:
```diff
@@ -0,0 +1,129 @@
+        return format_string
+
+    def step(self, epoch=None):
+        # Raise a warning if old pattern is detected
```
Since we do not have the old pattern, can we simplify the logic here?

raghuramank100 commented on torch/ao/sparsity/scheduler/base_scheduler.py:
```diff
@@ -0,0 +1,129 @@
+        # https://github.com/pytorch/pytorch/issues/20124
+        self._step_count += 1
+
+        class _enable_get_sl_call:
```
Not sure if this context manager is needed.

z-a-f replied:
Basically, this context manager is a guardrail to prevent the users from calling the `get_sl` manually. The reason I wouldn't want them to is because there is a separate function `get_last_sl`, which shows the latest computed sl. Otherwise, the sl will be reported incorrectly, depending on which epoch you are calling it on.

raghuramank100 commented on torch/ao/sparsity/scheduler/base_scheduler.py:
```diff
@@ -0,0 +1,129 @@
+        self.base_sl = [group['sparsity_level'] for group in sparsifier.module_groups]
+        self.last_epoch = last_epoch
+
+        # Following https://github.com/pytorch/pytorch/issues/20124
```
The logic here is hard to follow. Is this the only way to implement this?

z-a-f replied:
I put as many comments in here. But in general, I agree this is a little convoluted. I will change the first comment to make sure it is more clear.

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

raghuramank100 commented on test/ao/sparsity/test_scheduler.py:
```diff
@@ -0,0 +1,40 @@
+            return list(self.base_sl)
+
+
+class TestBaseScheduler(TestCase):
```
Please add tests for verifying (1) Sparsifier is called before scheduler. (2) get_sl method behaves as expected.

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

Commit:
Update on "[sparsity] Base sparsity level scheduler class"


Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).
